### PR TITLE
[OptionsResolver] Trigger deprecation only if the option is used

### PR DESCRIPTION
**Pull request by yceruto**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #28848
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/10496

It's true that showing a deprecation message when the option is not used is a bit annoying and it would be heavy to get rid of it.

Now, a deprecated option is triggered only when it's provided by the user or each time is being called from a lazy evaluation (except for deprecations based on the value, they're triggered only when provided by the user).

### CODE REVIEW

**Comment by yceruto** [comment]
There is still a desicion to make here to complete this patch:

> https://github.com/symfony/symfony/issues/28848#issuecomment-429877228 ... (but there should be a way to opt-out of the deprecation to be able to write a BC layer)

> https://github.com/symfony/symfony/pull/28860#discussion_r225246950 `offsetGet($option/*, bool $triggerDeprecation = true*/)` seems fine? That be the easiest :)

WDYT about the @ro0NL's idea? thoughts?

**Comment by nicolas-grekas** [comment]
> offsetGet($option/*, bool $triggerDeprecation = true*/)

looks like this is the way to go, isn't it?

**Comment by yceruto** [comment]
It should be ready now.

**Review by ro0NL — APPROVED** [review]

**Review by ro0NL — APPROVED** [review]

**Review by dunglas — APPROVED** [review]

File: src/Symfony/Component/OptionsResolver/OptionsResolver.php [hunk 1]
```diff
@@ -784,14 +790,20 @@ public function resolve(array $options = array())
...
     {
         if (!$this->locked) {
             throw new AccessException('Array access is only supported within closures of lazy options and normalizers.');
         }
 
+        $triggerDeprecation = 1 === \func_num_args() || \func_get_arg(1);
+
         // Shortcut for resolved options
         if (array_key_exists($option, $this->resolved)) {
+            if ($triggerDeprecation && isset($this->deprecated[$option]) && \is_string($this->deprecated[$option])) {
```
Comment (xabbuh) [inline src/Symfony/Component/OptionsResolver/OptionsResolver.php]: What happens when the deprecation is configured as a callback?

File: src/Symfony/Component/OptionsResolver/Options.php [hunk 2]
```diff
@@ -16,6 +16,8 @@
  *
  * @author Bernhard Schussek <bschussek@gmail.com>
  * @author Tobias Schultze <http://tobion.de>
+ *
+ * @method mixed offsetGet(string $option, bool $triggerDeprecation = true)
```
[the remark this answers is not in the document]
Comment (yceruto) [reply src/Symfony/Component/OptionsResolver/Options.php]: This method belong to the parent interface, implementing it should be mandatory, isn't it? This is about documenting the second argument for IDE detection when this interface is being used.

[the remark this answers is not in the document]
Comment (nicolas-grekas) [reply src/Symfony/Component/OptionsResolver/Options.php]: OK, works for me, I missed the parent interface :)

File: src/Symfony/Component/OptionsResolver/OptionsResolver.php [hunk 3]
```diff
@@ -784,14 +790,20 @@ public function resolve(array $options = array())
      * @throws OptionDefinitionException If there is a cyclic dependency between
      *                                   lazy options and/or normalizers
      */
-    public function offsetGet($option)
+    public function offsetGet($option/*, bool $triggerDeprecation = true*/)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/OptionsResolver/OptionsResolver.php]: the param should be duplicated in the docblock above (same as in the interface)

File: src/Symfony/Component/OptionsResolver/OptionsResolver.php [hunk 4]
```diff
@@ -784,14 +790,20 @@ public function resolve(array $options = array())
...
     {
         if (!$this->locked) {
             throw new AccessException('Array access is only supported within closures of lazy options and normalizers.');
         }
 
+        $triggerDeprecation = 1 === \func_num_args() || \func_get_arg(1);
```
[the remark this answers is not in the document]
Comment (yceruto) [reply src/Symfony/Component/OptionsResolver/OptionsResolver.php]: Sorry I don't get it, I copied the same logic from here https://github.com/symfony/symfony/blob/4964ffc8c052a592ff3f0fbeefffff65b54b75b4/src/Symfony/Component/HttpKernel/Kernel.php#L284 could you please tell me what's the difference in this case?

File: src/Symfony/Component/OptionsResolver/OptionsResolver.php (hunk 1 above)
Comment (yceruto) [reply src/Symfony/Component/OptionsResolver/OptionsResolver.php]: Exactly, deprecation based on the value should be triggers only if the option is provided by the user. See test cases.

**Review by nicolas-grekas — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Review by xabbuh — APPROVED** [review]

**Pull request merged**